annotation/tests/test_tasks_services.py

```diff
@@ -585,3 +636,1697 @@ def test_filter_tasks_db_no_files_or_jobs(
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
+    mock_stats_input.event_type = "closed"
+    with patch(
+        "annotation.tasks.services.get_task_stats_by_id", return_value=None
+    ) as mock_get_task_stats_by_id:
+
+        with pytest.raises(CheckFieldError):
+            add_task_stats_record(mock_session, task_id, mock_stats_input)
+
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        mock_session.add.assert_not_called()
+        mock_session.commit.assert_not_called()
+
+
+def test_add_task_stats_record_setattr(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
+    mock_stats_input.dict.return_value = {
+        "field1": "new_value1",
+        "field2": "new_value2",
+    }
+    mock_stats_db = MagicMock(spec=AnnotationStatistics)
+    mock_stats_db.field1 = "old_value1"
+    mock_stats_db.field2 = "old_value2"
+    with patch(
+        "annotation.tasks.services.get_task_stats_by_id",
+        return_value=mock_stats_db,
+    ) as mock_get_task_stats_by_id:
+        result = add_task_stats_record(mock_session, task_id, mock_stats_input)
+
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        assert mock_stats_db.field1 == "new_value1"
+        assert mock_stats_db.field2 == "new_value2"
+        assert mock_stats_db.updated is not None
+
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+
+        assert result == mock_stats_db
+
+
+def test_add_task_stats_record_create_new(mock_session: Mock):
+    task_id = 1
+    mock_stats_input = MagicMock(spec=AnnotationStatisticsInputSchema)
+    mock_stats_input.event_type = "open"
+    mock_stats_input.dict.return_value = {
+        "field1": "value1",
+        "field2": "value2",
+    }
+    mock_stats_db = MagicMock()
+    with patch(
+        "annotation.tasks.services.get_task_stats_by_id", return_value=None
+    ) as mock_get_task_stats_by_id, patch(
+        "annotation.tasks.services.AnnotationStatistics",
+        return_value=mock_stats_db,
+    ) as mock_annotation_statistics:
+        result = add_task_stats_record(mock_session, task_id, mock_stats_input)
+
+        mock_get_task_stats_by_id.assert_called_once_with(
+            mock_session, task_id
+        )
+        mock_annotation_statistics.assert_called_once_with(
+            task_id=task_id, **mock_stats_input.dict.return_value
+        )
+        mock_session.add.assert_called_once_with(mock_stats_db)
+        mock_session.commit.assert_called_once()
+
+        assert result == mock_stats_db
+
+
+def test_get_from_cache():
```

Comment:
too complex. check out this video, might help
https://www.youtube.com/watch?v=ujRo8n0LsU4